Modules/mathmodule.c

```diff
@@ -3024,49 +3024,55 @@ math_comb_impl(PyObject *module, PyObject *n, PyObject *k)
     long long i, factors;
 
     n = PyNumber_Index(n);
-    if (n == NULL)
+    if (n == NULL) {
         return NULL;
+    }
     k = PyNumber_Index(k);
     if (k == NULL) {
         Py_DECREF(n);
         return NULL;
     }
 
-    cmp = PyObject_RichCompareBool(n, k, Py_LT);
-    if (cmp < 0) {
-        goto error;
-    }
-    else if (cmp > 0) {
+    if (Py_SIZE(n) < 0) {
         PyErr_SetString(PyExc_ValueError,
-                        "n must be an integer greater than or equal to k");
-        goto error;
-    }
-
-    /* k = min(k, n - k) */
-    temp = PyNumber_Subtract(n, k);
-    if (temp == NULL) {
+                        "n must be a non-negative integer");
         goto error;
     }
-    cmp = PyObject_RichCompareBool(k, temp, Py_GT);
-    if (cmp > 0) {
-        Py_SETREF(k, temp);
-    }
-    else {
-        Py_DECREF(temp);
-        if (cmp < 0 && PyErr_Occurred()) {
+    if (Py_SIZE(n) <= Py_SIZE(k)) {
+        /* k = min(k, n - k) */
+        temp = PyNumber_Subtract(n, k);
+        if (temp == NULL) {
             goto error;
         }
+        if (Py_SIZE(temp) < 0) {
+            Py_DECREF(temp);
+            PyErr_SetString(PyExc_ValueError,
+                            "k must be an integer less than or equal to n");
+            goto error;
+        }
+        cmp = PyObject_RichCompareBool(k, temp, Py_GT);
+        if (cmp > 0) {
+            Py_SETREF(k, temp);
+        }
+        else {
+            Py_DECREF(temp);
+            if (cmp < 0 && PyErr_Occurred()) {
```

Comment:
I don't think we need the `PyErr_Occurred()` here. (Or do we?) As far as I know, if `PyObject_RichCompareBool` returns a negative value it's already guaranteed that there's an error set.

Reply:
Oh, I missed this!